src/kvstore/RocksEngine.cpp

```diff
@@ -461,5 +433,69 @@ ResultCode RocksEngine::createCheckpoint(const std::string& name) {
     return ResultCode::SUCCEEDED;
 }
 
-}  // namespace kvstore
-}  // namespace nebula
+ErrorOr<ResultCode, std::string> RocksEngine::backupTable(
+    const std::string& name,
+    const std::string& tablePrefix,
+    std::function<bool(const folly::StringPiece& key)> filter) {
+    auto backupPath = folly::stringPrintf(
+        "%s/checkpoints/%s/%s.sst", dataPath_.c_str(), name.c_str(), tablePrefix.c_str());
+    VLOG(3) << "Start writing the sst file with table (" << tablePrefix
+            << ") to file: " << backupPath;
+
+    auto parent = backupPath.substr(0, backupPath.rfind('/'));
+    if (!FileUtils::exist(parent)) {
+        if (!FileUtils::makeDir(parent)) {
+            LOG(ERROR) << "Make dir " << parent << " failed";
+            return ResultCode::ERR_BACKUP_EXISTED;
+        }
+    }
+
+    rocksdb::Options options;
+    rocksdb::SstFileWriter sstFileWriter(rocksdb::EnvOptions(), options);
+
+    std::unique_ptr<KVIterator> iter;
+    auto ret = prefix(tablePrefix, &iter);
+    if (ret != kvstore::ResultCode::SUCCEEDED) {
+        return ResultCode::ERR_BACKUP_EMPTY_TABLE;
+    }
+
+    if (!iter->valid()) {
+        return ResultCode::ERR_BACKUP_EMPTY_TABLE;
+    }
+
+    auto s = sstFileWriter.Open(backupPath);
+    if (!s.ok()) {
+        LOG(ERROR) << "BackupTable failed, path: " << backupPath << ", error: " << s.ToString();
+        return ResultCode::ERR_BACKUP_OPEN_FAILED;
+    }
+
+    while (iter->valid()) {
+        if (filter && filter(iter->key())) {
+            iter->next();
+            continue;
+        }
+        s = sstFileWriter.Put(iter->key().toString(), iter->val().toString());
+        if (!s.ok()) {
+            LOG(ERROR) << "BackupTable failed, path: " << backupPath << ", error: " << s.ToString();
+            sstFileWriter.Finish();
+            return ResultCode::ERR_BACKUP_TABLE_FAILED;
+        }
+        iter->next();
+    }
+
+    s = sstFileWriter.Finish();
+    if (!s.ok()) {
+        LOG(WARNING) << "Failure to insert data when backupTable,  " << backupPath
+                     << ", error: " << s.ToString();
+        return ResultCode::ERR_BACKUP_EMPTY_TABLE;
+    }
+
+    if (sstFileWriter.FileSize() == 0) {
+        return ResultCode::ERR_BACKUP_EMPTY_TABLE;
+    }
```

Comment:
Are empty tables allowed? for example roles table.

Reply:
Here the sstfilewriter is empty which means that the table does not exist.